core/google/cloud/iterator.py

```diff
@@ -45,40 +45,85 @@ def get_items_from_response(self, response):
 """
 
 
+import six
+
+
 class Iterator(object):
     """A generic class for iterating through Cloud JSON APIs list responses.
 
     :type client: :class:`google.cloud.client.Client`
     :param client: The client, which owns a connection to make requests.
 
-    :type path: string
+    :type path: str
     :param path: The path to query for the list of items.
 
+    :type page_token: str
+    :param page_token: (Optional) A token identifying a page in a result set.
+
+    :type max_results: int
+    :param max_results: (Optional) The maximum number of results to fetch.
+
     :type extra_params: dict or None
     :param extra_params: Extra query string parameters for the API call.
     """
 
     PAGE_TOKEN = 'pageToken'
-    RESERVED_PARAMS = frozenset([PAGE_TOKEN])
+    MAX_RESULTS = 'maxResults'
+    RESERVED_PARAMS = frozenset([PAGE_TOKEN, MAX_RESULTS])
 
-    def __init__(self, client, path, extra_params=None):
+    def __init__(self, client, path, page_token=None,
+                 max_results=None, extra_params=None):
         self.client = client
         self.path = path
         self.page_number = 0
-        self.next_page_token = None
+        self.next_page_token = page_token
+        self.max_results = max_results
+        self.num_results = 0
         self.extra_params = extra_params or {}
         reserved_in_use = self.RESERVED_PARAMS.intersection(
             self.extra_params)
         if reserved_in_use:
             raise ValueError(('Using a reserved parameter',
                               reserved_in_use))
+        self._curr_items = iter(())
 
     def __iter__(self):
-        """Iterate through the list of items."""
-        while self.has_next_page():
+        """The :class:`Iterator` is an iterator."""
+        return self
+
+    def _update_items(self):
+        """Replace the current items iterator.
+
+        Intended to be used when the current items iterator is exhausted.
+
+        After replacing the iterator, consumes the first value to make sure
+        it is valid.
+
+        :rtype: object
+        :returns: The first item in the next iterator.
+        :raises: :class:`~exceptions.StopIteration` if there is no next page.
+        """
+        if self.has_next_page():
             response = self.get_next_page_response()
-            for item in self.get_items_from_response(response):
-                yield item
+            items = self.get_items_from_response(response)
+            self._curr_items = iter(items)
+            return six.next(self._curr_items)
+        else:
+            raise StopIteration
+
+    def next(self):
+        """Get the next value in the iterator."""
+        try:
+            item = six.next(self._curr_items)
+            self.num_results += 1
+            return item
+        except StopIteration:
+            item = self._update_items()
+            self.num_results += 1
+            return item
```

Comment:
Not sure what you mean by "doing it this way". Could you be more specific?

RE: "could there be `has_more_items()`" --> it already exists and is used in this PR, called `has_next_page()`. Or maybe you mean something else?

----

I realized I could move the `self.num_results += 1` and `return item` to after the `try` / `except` block to avoid repeating myself. So I folded that change into the original commit while rebasing.